### PR TITLE
refactor doPurchaseUri

### PR DESCRIPTION
**Pull request by akinwale**

### CODE REVIEW

File: src/redux/reducers/file.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
+import * as ACTIONS from 'constants/action_types';
+import { buildURI, parseURI } from 'lbryURI';
+
+const reducers = {};
+const defaultState = {
```
Comment (seanyesmunt) [inline src/redux/reducers/file.js]: Please create a type for `FileState`

File: src/redux/actions/file.js [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+    // set save_file argument to True to save the file (old behaviour)
+    Lbry.get({ uri, save_file: saveFile })
+      .then(streamInfo => {
+        const timeout =
+          streamInfo === null || typeof streamInfo !== 'object' || streamInfo.error === 'Timeout';
+
+        if (timeout) {
+          dispatch({
+            type: ACTIONS.LOADING_FILE_FAILED,
```
Comment (seanyesmunt) [inline src/redux/actions/file.js]: `LOADING_FILE_X` is everywhere that `PURCHASE_URI_X` is. Do we need both?

File: src/redux/actions/file.js [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    const balance = selectBalance(state);
+    const fileInfo = makeSelectFileInfoForUri(uri)(state);
+    const downloadingByOutpoint = selectDownloadingByOutpoint(state);
+    const alreadyDownloading = fileInfo && !!downloadingByOutpoint[fileInfo.outpoint];
+    const alreadyStreaming = makeSelectStreamingUrlForUri(uri)(state);
+
+    if (alreadyDownloading || alreadyStreaming) {
+      dispatch({
+        type: ACTIONS.PURCHASE_URI_FAILED,
+        data: { uri, error: `Already downloading / streaming uri: ${uri}` },
```
Comment (kauffj) [inline src/redux/actions/file.js]: If this error surfaces to users, it should be written more clearly (Maybe just "Already fetching ${uri}")

File: src/redux/reducers/file.js [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+  return {
+    ...state,
+    failedPurchaseUris: newFailedPurchaseUris,
+    purchasedUris: newPurchasedUris,
+    purchasedStreamingUrls: newPurchasedStreamingUrls,
+  };
+};
+
+reducers[ACTIONS.PURCHASE_URI_FAILED] = (state: FileState, action: PurchaseUriFailed) => {
```
Comment (seanyesmunt) [inline src/redux/reducers/file.js]: This returns `FileState`

**Pull request merged**